### ISSUE TITLE
Blazor WebAssembly Component CSS not applied when running the published application!

### ISSUE THREAD

**Issue by DumboJetEngine** [dotnet/aspnetcore#35901]
For some reason, for one of my Blazor projects, the CSS for the MainLayout component (or at least some of it) is not applied to the published application. But it does during debugging!
One CSS rule that is obviously not applied is for the `.sidebar` element.
I see that the CSS file (******.Client.styles.css) has this selector: `.sidebar[b-divo65yscn]`
...while actual the element on the DOM is this: `<div class="sidebar" b-qhf4iwdnlz="">`
I am not sure if I am doing something wrong, but it seems weird that the attribute doesn't match.
There is only one sidebar in the entire application and the same code works fine while debugging.

This was my first serious attempt on using Blazor.
I wanted to create a portable webserver application that serves the WebAssembly code for the browser application to run and then exchange data with the server.
It is a simple and unfinished application, and I am sure the code appears to be written by a Blazor-newbie (myself !), but if it is possible to share its code with limited people (in case you want to spot the root of the problem in the source code), it would be nice. I have tried reproducing it in a brand new project, but I couldn't.

Hopefully, this is not something silly I did.
But it is weird for it to work fine on debug.

----

Some general feedback for Blazor (not sure if there is a better place to put it, or if anyone is interested for it):

- I generally like it a lot. But I think it can be improved a lot too.
- Sometimes the razor files get damaged (compiler error texts get dumped into them (!), code gets removed by autocomplete, code misaligned during auto-formatting, etc).
- When browser link builds the code in the background and a compilation error exists, the Error panel does not pop up (as far as I remember), preventing you from realizing there is an error and that the client code is not updated.
- We need a project template of Blazor WebAssembly using cookies for authentication.
- Also, some more advanced demo/template projects, demonstrating more features (accessing local storage without using JS, accessing cookies, setting focus on buttons without using JS, using Bootstrap modals without using JS, more advanced Blazor code, etc). Not sure if something like this exists in some corner of the internet.
- Also, a way to reference component instances in C# and expose an API from components to the outside world would be nice. So, that you can instruct a component by name to do something.
- I think that Razor components should have an OnDestroy method. Without this, de-registering events (event +=, -= operators) is an accumulating nightmare...
- I don't know if there is a security consideration to this, but allowing components to get their template as an injected parameter (as in AngularJS) might be cool. E.g. a component that has a fixed @code section, but the HTML template is injected on the fly. Or, specifying sub-templates for a component. E.g. the template of the node for a treeview component.
- The C# code highlighting (background highlight) inside the @code{} section is annoying. I can only see the value of this in the HTML template. Not outside of it.
- I got a bit disappointed by the limited entries in the list of globalization-cultures and timezones, but there is probably a file transfer size issue with adding more of those?
- FontIcons for cut/copy/rename/gear(settings) are missing. Not sure if you can do something about this or you are at the mercy of a 3rd party.
- AngularJS had a "bind-once" feature to improve performance. This updated the view based on the model only once, and then stopped checking for model changes. Is there something similar in Blazor? It would be super nice to have this in order to be able to boost the performance.
- Often, the code editor for razor files is filled with red curly underlines and reporting errors, while the build succeeds. Not sure how easy it would be to fix this. It is certainly annoying.

**Comment by javiercn** [comment]
@DumboJetEngine thanks for contacting us.

What version of Blazor/.NET where you using? The issue you've experienced might have been due to a caching issue, we would suggest trying to navigate with in-private mode or do a hard-refresh on the browser to see if the file being loaded is out of date.

> * Sometimes the razor files get damaged (compiler error texts get dumped into them (!), code gets removed by autocomplete, code misaligned during auto-formatting, etc).
> * The C# code highlighting (background highlight) inside the @code{} section is annoying. I can only see the value of this in the HTML template. Not outside of it.
> * Often, the code editor for razor files is filled with red curly underlines and reporting errors, while the build succeeds. Not sure how easy it would be to fix this. It is certainly annoying.
> * When browser link builds the code in the background and a compilation error exists, the Error panel does not pop up (as far as I remember), preventing you from realizing there is an error and that the client code is not updated.

I would suggest filing separate issues for these with details through the report a problem dialog as described [here](https://docs.microsoft.com/en-us/visualstudio/ide/how-to-report-a-problem-with-visual-studio?view=vs-2019)

> * We need a project template of Blazor WebAssembly using cookies for authentication.
We don't plan to do this in the near future, our current recommendation is to use JWTs (as that's what we ship on the templates).

> * Also, some more advanced demo/template projects, demonstrating more features (accessing local storage without using JS, accessing cookies, setting focus on buttons without using JS, using Bootstrap modals without using JS, more advanced Blazor code, etc). Not sure if something like this exists in some corner of the internet.

I would checkout https://github.com/dotnet-presentations/blazor-workshop as an example of a more complete app that uses many features.

> * Also, a way to reference component instances in C# and expose an API from components to the outside world would be nice. So, that you can instruct a component by name to do something.

I would checkout https://docs.microsoft.com/en-us/aspnet/core/blazor/components/?view=aspnetcore-5.0#capture-references-to-components-1

> * I think that Razor components should have an OnDestroy method. Without this, de-registering events (event +=, -= operators) is an accumulating nightmare...

We recommend you implement IDisposable to do this, it's not very common to use many event handlers in Blazor applications.
If you find yourself doing this a lot, you can put the code in a base class and provide your own "OnDestroy" method where you can register callbacks to be called during Dispose.

> * I don't know if there is a security consideration to this, but allowing components to get their template as an injected parameter (as in AngularJS) might be cool. E.g. a component that has a fixed @code section, but the HTML template is injected on the fly. Or, specifying sub-templates for a component. E.g. the template of the node for a treeview component.

Blazor does this automatically by converting the contents of a component tag into a RenderFragment callback you can use to render the contents wherever you want. See [here](https://docs.microsoft.com/en-us/aspnet/core/blazor/components/?view=aspnetcore-5.0#razor-templates-1) for details

> * I got a bit disappointed by the limited entries in the list of globalization-cultures and timezones, but there is probably a file transfer size issue with adding more of those?

Not sure about this one, I believe there is a way to even load all the cultures. @pranavkm might know more.

> * FontIcons for cut/copy/rename/gear(settings) are missing. Not sure if you can do something about this or you are at the mercy of a 3rd party.
This is not something that we control other than we ship it in the default template, you are free to replace it with any other icon library you want.

> * AngularJS had a "bind-once" feature to improve performance. This updated the view based on the model only once, and then stopped checking for model changes. Is there something similar in Blazor? It would be super nice to have this in order to be able to boost the performance.

You don't need this in Blazor, if you only want to "bind once", you only need to set the `value` attribute instead of using `@bind`

**Comment by DumboJetEngine** [comment]
@javiercn 
Thank you for your reply.

I have tried the in-private mode and clearing the cache in my browser, but it didn't seem to help.
I target .NET 5 and my project references the nuget package `Microsoft.AspNetCore.Components.WebAssembly` version `5.0.7`.
So, I guess `5.0.7` is the Blazor version, correct?
I just saw there is a newer version available (`5.0.9`), so I have tried that too. I still get the same problem.

> Not sure about this one, I believe there is a way to even load all the cultures. @pranavkm might know more.

I've noticed that there are only some deprecated options to do this (and load OS cultures), but I have the impression they don't work already.

> You don't need this in Blazor, if you only want to "bind once", you only need to set the value attribute instead of using @bind

I was mainly thinking about non-input elements, say a span tag that has its content set like so:
`<span>@contents</span>`
I am not very familiar on how Blazor works internally, but I was guessing that every time it checks for changes, the existence of this code alone would trigger a check on whether the value has changed in order to update the tag contents. This extra CPU cost is what I am trying to avoid. Because I've noticed that there is delay after certain amount of page content gets rendered, just like when using AngularJS (I render a table with many rows).

Thank you for the links! I will take a look. :)

> We recommend you implement IDisposable to do this, it's not very common to use many event handlers in Blazor applications.

I had to use an event in a razor component to make the routing work the way I wanted. Essentially, I wanted to capture the `path` variable in the below component, every time it changed (either by code or by user navigation):
```
@page "/explorer/{path?}"
...
navigationManager.LocationChanged += OnLocationChanged;
```
I will look in your samples to see if I can find a better way... Probably I will.

**Comment by javiercn** [comment]
> I have tried the in-private mode and clearing the cache in my browser, but it didn't seem to help.
> I target .NET 5 and my project references the nuget package `Microsoft.AspNetCore.Components.WebAssembly` version `5.0.7`.
> So, I guess `5.0.7` is the Blazor version, correct?
> I just saw there is a newer version available (`5.0.9`), so I have tried that too. I still get the same problem.
> What I do is to publish the server project, and the client project (that is referenced from the server project) gets published with it. Not sure if Blazor needs any special treatment compared to other types of projects.

All that you are describing should work.

In order for us to continue investigating we would need a minimal repro project as a public github repo that we can try things with. We haven't seen this reported by any other person, so its not clear what might be going on or if it might be specific to your machine/environment

Another thing you could do is to try the .NET 6.0 SDK (you can grab a nightly build from dotnet/installer)

> I was mainly thinking about non-input elements, say a span tag that has its content set like so:
<span>@contents</span>
I am not very familiar on how Blazor works internally, but I was guessing that every time it checks for changes, the existence of this code alone would trigger a check on whether the value has changed in order to update the tag contents. This extra CPU cost is what I am trying to avoid. Because I've noticed that there is delay after certain amount of page content gets rendered, just like when using AngularJS (I render a table with many rows).

Blazor works more like other frameworks like Vue or React, where components explicitly decide when to render. A re-render is only triggered when a component receives new parameters or as a result of an event being triggered. You can override `ShouldRender` to completely control when a component re-renders.

**Comment by DumboJet** [comment]
OK, I will attempt to create a stripped-down version of my code pretty soon.

**Comment by DumboJetEngine** [comment]
OK, I will attempt to create a stripped-down version of my code pretty soon.

**Comment by DumboJetEngine** [comment]
Hello again.

Regarding the "CSS not applied" bug, here is a demo project (creating it was a bit painful, that's why it took me some time):
https://github.com/DumboJetEngine/this-is-a-test/tree/master
To run it you will need a pfx certificate inside the `Server/Cert` folder.
When debugging from VS everything seems normal, while when running the published version with these settings:
![image](https://user-images.githubusercontent.com/8676903/133694490-52df4048-0af5-4816-a90d-520d41375bc7.png)
...it gives you this broken page:
![image](https://user-images.githubusercontent.com/8676903/133694550-b508967e-f2c3-493d-940e-680caf19de9b.png)
(the sidebar is completely missing)
Let me know if you can reproduce this.

**Comment by G31st** [comment]
Hi all, has there been any fix to this, surely this cannot be broken for everyone ? Locally all styles work perfectly and I've updated to latest stable on all packages.  I've also tried a few different publish profile settings mostly focused around the runtimeidentifier, but with no lucky

  <ItemGroup>
    <PackageReference Include="Blazored.LocalStorage" Version="4.1.5" />
    <PackageReference Include="Blazorise.Bootstrap" Version="0.9.4.4" />
    <PackageReference Include="Blazorise.Icons.FontAwesome" Version="0.9.4.4" />
    <PackageReference Include="Blazorise.Icons.Material" Version="0.9.4.4" />
    <PackageReference Include="Blazorise.TreeView" Version="0.9.4.4" />
    <PackageReference Include="Microsoft.AspNetCore.Components.WebAssembly" Version="5.0.10" />
    <PackageReference Include="Microsoft.AspNetCore.Components.WebAssembly.Authentication" Version="5.0.10" />
    <PackageReference Include="Microsoft.AspNetCore.Components.WebAssembly.DevServer" Version="5.0.10" PrivateAssets="all" />
    <PackageReference Include="Microsoft.Extensions.Http" Version="5.0.0" />
    <PackageReference Include="System.Net.Http.Json" Version="5.0.0" />
    <PackageReference Include="YamlDotNet" Version="11.2.1" />
  </ItemGroup>


Publish Settings : 

 <PropertyGroup>
    <DeleteExistingFiles>True</DeleteExistingFiles>
    <ExcludeApp_Data>False</ExcludeApp_Data>
    <LaunchSiteAfterPublish>True</LaunchSiteAfterPublish>
    <LastUsedBuildConfiguration>Release</LastUsedBuildConfiguration>
    <LastUsedPlatform>Any CPU</LastUsedPlatform>
    <PublishProvider>FileSystem</PublishProvider>
    <PublishUrl>..\Publish\www-root</PublishUrl>
    <WebPublishMethod>FileSystem</WebPublishMethod>
    <SiteUrlToLaunchAfterPublish />
    <TargetFramework>net5.0</TargetFramework>
    <SelfContained>true</SelfContained>
    <RuntimeIdentifier>browser-wasm</RuntimeIdentifier>
    <PublishTrimmed>True</PublishTrimmed>
    <PublishSingleFile>False</PublishSingleFile>
  </PropertyGroup>

**Comment by javiercn** [comment]
@G31st Using your sample app the UI loads fine for me.

![image](https://user-images.githubusercontent.com/6995051/135436844-a354268a-4b2b-4829-aaf9-bd113db34254.png)

**Comment by javiercn** [comment]
@G31st can you check on the network tab in the browser to see if there is any failure loading any css file?

**Comment by nic547** [comment]
I had the same issue, publishing a self-contained application breaking the css isolation feature. The issue appeared on edge/chromium/Firefox and the loaded stylesheets matched the file on disk. I tried using VS, dotnet and dotnet via WSL2 for publishing and none worked. 

I also found a workaround for the issue, running dotnet publish on a second clone of my repo did work, matching the experience that the issue cannot be simply reproduced by cloning and publishing a project. Can anyone else observe the same behavior?

**Comment by DumboJetEngine** [comment]
For some reason, I stopped being able to reproduce the problem on the demo project.

So, I returned to the original project to test.
There, the problem was remaining.
Rebuilding and republishing did not solve the problem: the published app had its CSS files broken (using wrong attributes, e.g. for the `.sidebar` element).
To verify that I get the CSS files from the disk, I copied the attribute from the `.sidebar` element from browser's DOM and pasted it in the CSS file on disk, replacing the wrong value that was in there. After reloading with Ctrl+F5, the sidebar appeared (but some remaining stuff was broken, as expected).
Then I tried deleting all bin/obj folders. I deleted the ones in the Server project and the ones in the Client project.
Then, I rebuilt and republished.
The result: everything got fixed. The layout was correct everywhere!

I tried a bit to break it again from within VS, but with not much luck.
Intentionally breaking the CSS file on disk did not provide a permanent break. The republish operation updated the file with the correct values.
As a summary, it appears that the build process, mistakenly, doesn't copy the CSS file under some conditions or it copies the wrong (out of date) file.
Deleting the bin/obj folders fixes the problem, but rebuilding and republishing doesn't.
Even when the option for `delete all existing files prior to built` is selected on the publish profile settings.

So, I don't know...
I hope the above is helpful to someone.
And if someone can think of what the conditions that trigger this failure to update the CSS file can be, let me know.
Otherwise, I could try playing around the solution and see it the problem returns.
